### PR TITLE
feat(rust, python): add `drop constraint` operation

### PR DESCRIPTION
**Pull request by ion-elgreco**
# Description
Adds `drop constraint` as an operation, and also exposed it to python under the alter namespace 😄 

# Related Issue(s)
- closes https://github.com/delta-io/delta-rs/issues/2070

### CODE REVIEW

**Comment by ion-elgreco** [comment]
@hntd187 do you want to review this one?

File: crates/deltalake-core/src/operations/drop_constraints.rs [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+                )));
+            } else if !this.raise_if_not_exists && !found_constraint {
+                return Ok(DeltaTable::new_with_state(this.log_store, this.snapshot));
+            }
+
+            metadata
+                .configuration
+                .remove(format!("delta.constraints.{}", name).as_str());
+
+            let mut contains_constraints: bool = false;
```
Comment (hntd187) [inline crates/deltalake-core/src/operations/drop_constraints.rs:96]: you could instead write this as `let contains_constraints = metadata.configuration.keys().any(|k| k.starts_with("delta.constraints"));`

File: crates/deltalake-core/src/operations/drop_constraints.rs [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+                } else {
+                    continue;
+                }
+            }
+
+            let operational_parameters = HashMap::from_iter([("name".to_string(), json!(&name))]);
+
+            let operations = DeltaOperation::DropConstraint { name: name.clone() };
+
+            let app_metadata = match this.app_metadata {
```
Comment (hntd187) [inline crates/deltalake-core/src/operations/drop_constraints.rs:110]: `.unwrap_or_default()` instead

File: crates/deltalake-core/src/operations/drop_constraints.rs [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+                operation: Some(operations.name().to_string()),
+                operation_parameters: Some(operational_parameters),
+                read_version: Some(this.snapshot.version()),
+                isolation_level: Some(IsolationLevel::Serializable),
+                is_blind_append: Some(false),
+                info: app_metadata,
+                ..Default::default()
+            };
+
+            let protocol = if !contains_constraints {
```
Comment (hntd187) [inline crates/deltalake-core/src/operations/drop_constraints.rs:126]: Do you know if other implementations downgrade the protocol as a result of dropping the constraint? I can check this for you, but I don't think the protocol says anything about whether or not it's required to do this.

File: crates/deltalake-core/src/operations/drop_constraints.rs [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+                    reader_features: old_protocol.reader_features.clone(),
+                    writer_features: old_protocol.writer_features.clone(),
+                })
+            } else {
+                None
+            };
+
+            let mut actions = vec![Action::CommitInfo(commit_info), Action::Metadata(metadata)];
+
+            protocol.map(|protocol| actions.push(Action::Protocol(protocol)));
```
Comment (hntd187) [inline crates/deltalake-core/src/operations/drop_constraints.rs:149]: while it's more code a `if let Some(protocol) = protocol {` makes it clearer on what you are intending here.

**Review by hntd187 — CHANGES_REQUESTED** [review]

File: crates/deltalake-core/src/operations/drop_constraints.rs (hunk 3 above)
Comment (ion-elgreco) [reply crates/deltalake-core/src/operations/drop_constraints.rs:126]: I just assumed here it's safe to do since the protocol version was only for constraints, and if there are no constraints imho it should not be left there. 

You could check what spark-delta here does :) But it should be fine I think, I don't believe it violates the protocol

File: crates/deltalake-core/src/operations/drop_constraints.rs (hunk 2 above)
Comment (ion-elgreco) [reply crates/deltalake-core/src/operations/drop_constraints.rs:110]: Ah of course there is something for that :P Thanks!

File: crates/deltalake-core/src/operations/drop_constraints.rs (hunk 1 above)
Comment (ion-elgreco) [reply crates/deltalake-core/src/operations/drop_constraints.rs:96]: Ah didn't know this method existed.

File: crates/deltalake-core/src/operations/drop_constraints.rs (hunk 3 above)
Comment (ion-elgreco) [reply crates/deltalake-core/src/operations/drop_constraints.rs]: Discussed with @roeap on slack and it wouldn't be safe to do this, so I've removed it in ths operation, but we may look at adding an explicit operation to downgrade

File: crates/core/src/operations/drop_constraints.rs [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+                    "Constraint with name: {} doesn't exists",
+                    name
+                )));
+            } else if !this.raise_if_not_exists && !found_constraint {
+                return Ok(DeltaTable::new_with_state(this.log_store, this.snapshot));
+            }
+
+            metadata
+                .configuration
+                .remove(format!("delta.constraints.{}", name).as_str());
```
Comment (roeap) [inline crates/core/src/operations/drop_constraints.rs:94]: I think we can simplify this a bit by leveraging that remove will return the removed value if the key existed.

 ```suggestion
            if metadata.configuration.remove(&configuration_key).is_none() {
                if this.raise_if_not_exists {
                    return Err(DeltaTableError::Generic(format!(
                        "Constraint with name: {} doesn't exists",
                        name
                    )));
                }
                return Ok(DeltaTable::new_with_state(this.log_store, this.snapshot));
            }
```

File: crates/core/src/operations/drop_constraints.rs [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+            let contains_constraints = metadata
+                .configuration
+                .keys()
+                .any(|k| k.starts_with("delta.constraints"));
```
Comment (roeap) [inline crates/core/src/operations/drop_constraints.rs:99]: do we still need this?

File: python/Cargo.toml [hunk 7]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "deltalake-python"
-version = "0.15.2"
+version = "0.16.0"
```
Comment (roeap) [inline python/Cargo.toml:3]: Should we do version bump in a separate PR? Or rather should we come up with a strategy how we usually do this?

**Review by roeap — COMMENTED** [review]
looking good, left a few mostly style related comments.

File: crates/core/src/operations/drop_constraints.rs [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+    type IntoFuture = BoxFuture<'static, Self::Output>;
+
+    fn into_future(self) -> Self::IntoFuture {
+        let mut this = self;
+
+        Box::pin(async move {
+            let name = match this.name {
+                Some(v) => v,
+                None => return Err(DeltaTableError::Generic("No name provided".to_string())),
+            };
```
[the remark this answers is not in the document]
Comment (ion-elgreco) [reply crates/core/src/operations/drop_constraints.rs:77]: Ah there seems to be an inbuilt pattern for all these things already 😆

File: python/Cargo.toml (hunk 7 above)
Comment (ion-elgreco) [reply python/Cargo.toml:3]: Usually it's a separate one, but wanted to save time here in case this ended up in main this week : )

**Comment by ion-elgreco** [comment]
@roeap thanks for the feedback! I've resolved the comments

**Review by hntd187 — APPROVED** [review]
LGMT

File: crates/core/src/operations/mod.rs [hunk 9]
```diff
@@ -199,6 +201,12 @@ impl DeltaOps {
     pub fn add_constraint(self) -> ConstraintBuilder {
         ConstraintBuilder::new(self.0.log_store, self.0.state.unwrap())
     }
+
+    /// Drops constraints from a table
+    #[must_use]
```
Comment (rtyler) [inline crates/core/src/operations/mod.rs:206]:
```suggestion
   #[cfg(feature = "datafusion")]
    #[must_use]
```

This should be guarded on the `datafusion` feature

**Review by rtyler — COMMENTED** [review]
Small change and then we're ready to roll :guitar:

File: crates/core/src/operations/mod.rs (hunk 9 above)
Comment (ion-elgreco) [reply crates/core/src/operations/mod.rs:206]: Ah oops :)

**Review by rtyler — APPROVED** [review]
